### PR TITLE
[WIP] Adds performance build guide

### PR DESCRIPTION
**Pull request by twokul**
A good place for developers to start tracking issues down.

### CODE REVIEW

**Comment by jonnii** [comment]
##### Q

My windows builds are slow!

##### A

Buy a mac :trollface:

File: PERF_GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Build performance
+
+When we talk about build performance, it is important to understand that there
+are server build types:
```
Comment (csantero) [inline PERF_GUIDE.md]: server -> several?

File: PERF_GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# Build performance
+
+When we talk about build performance, it is important to understand that there
+are several build types:
+
++ cold build (booting your app up for the first time)
```
Comment (twokul) [inline PERF_GUIDE.md]: elaborate on differences between build types; set expectations on build times; if app doesn't meet expectations, there might be a problem; how do we identify the problem and what can tools help us with it;

File: PERF_GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+### `broccoli-viz`
+
+#### Visualization
+
+To visualize build tree, we use [graphviz](http://www.graphviz.org/). To install it run `brew install graphviz` or download it directly from [here](http://www.graphviz.org/Download.php).
+
+To generate visualization:
+
++ `BROCCOLI_VIZ=true ember build`
++ `dot -Tpng grap.<version>.dot > out.png` (each build, will generate an additional graph.<build-number>.dot  graph.<build-number>.json)
```
Comment (Dhaulagiri) [inline PERF_GUIDE.md]: missing the `h` in `grap.<version>`

**Comment by stefanpenner** [comment]
@twokul any more time to work on this?

File: PERF_GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+Cold build is the slowest because the cache is not yet populated and the application is booting for the first time.
+Build time varies and depends on the number of dependencies that application has but ballpark should be around 5 seconds
+for small to middle size applications and around 15 seconds for large size applications.
+
+Warm build is faster then cold one because the cache was populated already and it takes less time re-compute dependecies.
+Build time varies but ballpark should be around 2 seconds for small to middle size applications and around 10 seconds for
+large size applications.
+
+Rebuild is the fastest operantion because the cache is in prime. Build time varies but ballpark should be around 200-300ms
+for small to middle size applications and up to 1 second for large size applications.
```
Comment (MiguelMadero) [inline PERF_GUIDE.md]: what's considered a large application? Is 10 klocs or 150 klocs a large app?

Comment (stefanpenner) [inline PERF_GUIDE.md]: One of my apps is about 200kloc js app code, and it has sub second JS change times. 400ms-600ms (with source maps disabled) about 1s with sourceMaps on.

We have been working on a different sourceMap VLQ encoder/decoder which is about an order of magnitude faster. Hopefully I'll have a chance over the next week or so to integrate it into broccoli-concat.

Comment (MiguelMadero) [inline PERF_GUIDE.md]: Thanks that makes sense. I was just curious about what small, middle, large meant for this guide. This gives us a good reference, for a similar sized app with 1.13.12 we're at ~11secs for a rebuild, which is way better than with 1.13.8, but still slow enough to break the dev flow. I'll dig later this week to see where the problem is. My guess is that it's something specific to what we're doing.

Comment (stefanpenner) [inline PERF_GUIDE.md]: we typically see slower builds with:

1. old plugins, static-compiler/ old jshint old broccoli-filter etc.etc.etc.
2. anti-virus
3. strange custom plugins

Comment (stefanpenner) [inline PERF_GUIDE.md]: 1.13.12 is also without some pretty important internal performance work, please be sure to atleast be at 1.13.13.

**Comment by stefanpenner** [comment]
this is sufficiently useful to merge, we can continue to iterate as we have time.

**Pull request merged**